internal/handlers/handler_firstfactor.go

```diff
@@ -2,127 +2,200 @@ package handlers
 
 import (
 	"fmt"
+	"math/rand"
 	"time"
 
 	"github.com/authelia/authelia/internal/authentication"
+	"github.com/authelia/authelia/internal/configuration/schema"
+	"github.com/authelia/authelia/internal/logging"
 	"github.com/authelia/authelia/internal/middlewares"
 	"github.com/authelia/authelia/internal/regulation"
 	"github.com/authelia/authelia/internal/session"
+	"github.com/authelia/authelia/internal/utils"
 )
 
-// FirstFactorPost is the handler performing the first factory.
-//nolint:gocyclo // TODO: Consider refactoring time permitting.
-func FirstFactorPost(ctx *middlewares.AutheliaCtx) {
-	bodyJSON := firstFactorRequestBody{}
-	err := ctx.ParseBody(&bodyJSON)
+func getDelayAuthSettings(config schema.AuthenticationBackendConfiguration) (bool, time.Duration) {
+	log := logging.Logger()
+
+	if !config.DisableDelayAuth {
+		var duration time.Duration
+
+		rand.Seed(time.Now().UnixNano())
+
+		if config.File != nil {
+			algorithm, err := authentication.ConfigAlgoToCryptoAlgo(config.File.Password.Algorithm)
+			if err != nil {
+				panic(err)
+			}
+
+			password := utils.RandomString(20, authentication.HashingPossibleSaltCharacters)
+			start := time.Now()
+			_, _ = authentication.HashPassword(password, "",
+				algorithm, config.File.Password.Iterations,
+				config.File.Password.Memory*1024, config.File.Password.Parallelism,
+				config.File.Password.KeyLength, config.File.Password.SaltLength)
+			duration = time.Since(start) + (time.Duration(rand.Int31n(50) + 150))
+		}
+
+		if duration < 450*time.Millisecond {
+			duration = time.Duration(rand.Intn(50)+450) * time.Millisecond
+		}
+
+		log.Debugf("1FA authentication requests will not return to clients until %dms after the request was received "+
+			"to prevent username enumeration.", duration/time.Millisecond)
 
-	if err != nil {
-		handleAuthenticationUnauthorized(ctx, err, authenticationFailedMessage)
+		return true, duration
+	}
+
+	log.Warn("1FA authentication requests will not be delayed as it has been disabled by configuration option " +
+		"authentication_backend.disable_delay_auth, this reduces security and is not recommended.")
+
+	return false, 0 * time.Millisecond
+}
+
+func doDelayAuth(ctx *middlewares.AutheliaCtx, username string, receivedTime time.Time, enabled, successful bool, duration *time.Duration) {
+	if !enabled {
+		ctx.Logger.Warnf("Skipping the authentication delay of user %s since authentication delay is disabled by configuration", username)
 		return
 	}
 
-	bannedUntil, err := ctx.Providers.Regulator.Regulate(bodyJSON.Username)
+	// Automatically grow to the largest detected successful login.
+	if successful {
+		delay := time.Since(receivedTime)
+		if delay > *duration {
```

Comment:
duration will never decrease back after an incident in which latency has raised drastically.